controllers/idler/aap_idler.go

```diff
@@ -0,0 +1,265 @@
+				requeueAfter = shorterDuration(requeueAfter, killAfter)
+			}
+		}
+
+		// Check if we need to send a notification and proceed with the next pod
+		if idledAAPName != "" {
+			// The AAP has been Idled
+			// A notification should be sent
+			i.notifyUser(podCtx, idler, idledAAPName, "Ansible Automation Platform")
+
+			idledAAPs = append(idledAAPs, idledAAPName)
+			if len(idledAAPs) == len(running) {
+				// All AAPs are idled, no need to check the rest of the pods
+				// no need to schedule any aap-specific requeue
+				return 0, nil
+			}
+		}
+	}
+
+	// there is at least one aap instance, schedule the next reconcile
+	return requeueAfter, nil
+}
+
+// getRunningAAPs returns the list of all AAP CRs that are not idled from the namespace the idler was created for
+func (i *aapIdler) getRunningAAPs(ctx context.Context, idler *toolchainv1alpha1.Idler) ([]unstructured.Unstructured, error) {
+	// Check if there is any AAP CRs in the namespace
+	aapList, err := i.dynamicClient.Resource(*i.aapGVR).Namespace(idler.Name).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	running := make([]unstructured.Unstructured, 0, len(aapList.Items))
+	for _, aap := range aapList.Items {
+		// we don't need to check if the field was found - if not found, then the value is "false" (not idled)
+		idled, _, err := unstructured.NestedBool(aap.UnstructuredContent(), "spec", "idle_aap")
+		if err != nil {
+			return nil, err
+		}
+		if !idled {
+			running = append(running, aap)
+		}
+	}
+	return running, nil
+}
+
+const twoHours = 2 * 60 * 60 // in seconds
+
+func aapTimeoutSeconds(idlerTimeout int32) int32 {
+	// Check if the idler timeout is less than two hours and if so, set it to half of the timeout.
+	// Otherwise, subtract one hour from the timeout.
+	// This is to ensure that the AAP idler kicks in before the main idler.
+	timeoutSeconds := idlerTimeout
+	if timeoutSeconds <= twoHours {
+		timeoutSeconds = timeoutSeconds / 2
+	} else {
+		timeoutSeconds = timeoutSeconds - twoHours/2
+	}
+	return timeoutSeconds
+}
+
+// ensureAAPIdled checks if the long-running or crash-looping pod belongs to an AAP instance and if so, ensures that the AAP is idled.
+// Returns the AAP resource name in case it was idled, or an empty string if it was not idled.
+func (i *aapIdler) ensureAAPIdled(ctx context.Context, pod corev1.Pod, alreadyIdled []string) (string, error) {
+	aap, err := i.getAAPOwner(ctx, &pod)
+	if err != nil || aap == nil {
+		return "", err // either error or no AAP owner found
+	}
+	for _, idled := range alreadyIdled {
+		if aap.GetName() == idled {
+			// it is already idled - nothing to do
+			return "", nil
+		}
+	}
+
+	// Patch the aap resource by setting spec.idle_aap to true in order to idle it
+	patch := []byte(`{"spec":{"idle_aap":true}}`)
+	_, err = i.dynamicClient.Resource(*i.aapGVR).Namespace(pod.Namespace).Patch(ctx, aap.GetName(), types.MergePatchType, patch, metav1.PatchOptions{})
+	if err != nil {
+		return "", err
```

Comment:
I wonder if we actually want to return an error here... I'm just a bit concern about potential problems with AAP idling (changed API and/or some new webhook preventing us to patch it or something like this) which would generate an error and potentially (partially) break the entire idler... Maybe it would be safer to log the error but do not return it? WDYT?
I don't have any strong opinion.

Reply:
It's a good point. I was thinking about the same,  but I completely forgot to address it. 
At the first glance, any error returned from the AAP idler shouldn't affect the main-idler logic (as it's executed after it), but if the AAP idler keeps returning errors the controller, then it could negatively affect the reconciles scheduled by the main idler. 

addressed in c9725fd8
* it "ignores" errors within the idler (eg. when tracking the owners or trying to patch an AAP instance), simply to give a chance to idle the other pods/AAP instances. It only collects the errors and returns them at the end.
* it doesn't return the errors returned from the AAP idler - it only logs them and updates the Ready condition of the Idler CR